burstphoto/denoise.swift

```diff
@@ -319,32 +318,36 @@ func calculate_temporal_average(progress: ProcessingProgress, mosaic_pattern_wid
         }
     }
     
-    if (!uniform_exposure && white_level != -1 && black_level[0][0] != -1 && mosaic_pattern_width == 2) {
+    // if color_factor is NOT available, it is set to a negative value earlier in the pipeline
+    if (white_level != -1 && black_level[0][0] != -1 && color_factors[0][0] > 0 && mosaic_pattern_width == 2) {
         
         // initialize weight texture used for normalization of the final image
         let norm_texture_descriptor = MTLTextureDescriptor.texture2DDescriptor(pixelFormat: .r32Float, width: final_texture.width, height: final_texture.height, mipmapped: false)
         norm_texture_descriptor.usage = [.shaderRead, .shaderWrite]
         norm_texture_descriptor.storageMode = .private
         let norm_texture = device.makeTexture(descriptor: norm_texture_descriptor)!
         fill_with_zeros(norm_texture)
+        // initialize scalar weight used for normalization of the final image
+        var norm_scalar = 0
         
-        // temporal averaging with exposure weighting
+        // temporal averaging with extrapolation of highlights or exposure weighting
         for comp_idx in 0..<textures.count {
             let comp_texture = prepare_texture(textures[comp_idx], hotpixel_weight_texture, 0, 0, 0, 0, exposure_bias[exp_idx]-exposure_bias[comp_idx], black_level, comp_idx)
-            add_texture_exposure(comp_texture, final_texture, norm_texture, exposure_bias[comp_idx]-exposure_bias[exp_idx], ((comp_idx==exp_idx) ? 1_000_000 : white_level), black_level[comp_idx])
+                       
+            if exposure_bias[comp_idx] == exposure_bias[exp_idx] {
+                // for uniform exposure or for the darkest frame in an exposure bracketed burst: add frame and apply extrapolation of green channels for very bright pixels. All pixels in the frame get the same global weight of 1. Therefore a scalar value for normalization storing the sum of accumulated frames is sufficient.
+                add_texture_highlights(comp_texture, final_texture, 1, white_level, black_level[comp_idx], color_factors[comp_idx])
```

Comment:
Does `add_texture_highlights` still need the `n_texture` argument? This is now the only place where it's used.

Reply:
Good catch, thank you. Will remove the argument n_textures.